sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -442,8 +443,38 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
       buildCast[Int](_, d => null)
     case TimestampType =>
       buildCast[Long](_, t => timestampToLong(t))
-    case x: NumericType =>
-      b => x.numeric.asInstanceOf[Numeric[Any]].toLong(b)
+    case ByteType =>
+      b => b.asInstanceOf[Byte].toLong
+    case ShortType =>
+      b => b.asInstanceOf[Short].toLong
+    case IntegerType =>
+      b => b.asInstanceOf[Int].toLong
```

Comment:
We need the three entries above? `case x: NumericType` you removed in this pr is enough for those cases?

Reply:
Not sure if there is performance difference here. The motivation is that there is only there types here and by doing this might slightly improve the performance. @rednaxelafx @kiszk

Reply:
Ur, I see. Good idea to ask the JVM guys, hahaha

Reply:
In summary, I think that this slighly improve improve performance.

My guess (I have not seen generated code by HotSpot) is that this cast (e.g. `b.asInstanceOf[Int]`) removes table lookup of invocation of `toLong`. As a result, we expect the code in `toLong` is inlined to caller. On the other hand, the original code (`asInstanceOf[Numeric[Any]].toLong(b)`) has a table lookup for invoking `toLong`. Thus, it is not expected to apply inlining.